R/init.R

```diff
@@ -142,26 +143,13 @@ init <- function(data,
   if (length(landing) > 1L) stop("Only one `landing_popup_module` can be used.")
   modules <- drop_module(modules, "teal_module_landing")
 
-  # resolve modules datanames
-  datanames <- teal.data::get_dataname(data)
-  join_keys <- teal.data::get_join_keys(data)
-  modules <- resolve_modules_datanames(modules = modules, datanames = datanames, join_keys = join_keys)
-
-  if (!inherits(filter, "teal_slices")) {
-    checkmate::assert_subset(names(filter), choices = datanames)
-    # list_to_teal_slices is lifted from teal.slice package, see zzz.R
-    # This is a temporary measure and will be removed two release cycles from now (now meaning 0.13.0).
-    filter <- list_to_teal_slices(filter)
-  }
-  # convert teal.slice::teal_slices to teal::teal_slices
-  filter <- as.teal_slices(as.list(filter))
-
-  # Calculate app hash to ensure snapshot compatibility. See ?snapshot. Raw data must be extracted from environments.
+  # Calculate app hash to ensure snapshot compatibility.
+  # See ?snapshot. Raw data must be extracted from environments.
   hashables <- mget(c("data", "modules"))
   hashables$data <- if (inherits(hashables$data, "teal_data")) {
     as.list(hashables$data@env)
-  } else if (inherits(hashables$data, "ddl")) {
-    attr(hashables$data, "code")
+  } else if (inherits(data, "teal_data_module")) {
+    body(data$server)
```

Comment:
I don't quite get the idea behind `hashables` but it looks to be an unique identifier of the input data (e.g. you do calculate hash on env with data tables in it). Here you are extracting server function body and calculate hashes on this. Please note that the same server function might return different object (vide: `runif()` but more realisticaly here: module with file upload). I don't think this is what we expect in the next step.

(The same mistake below where we call `hashables$data$get_code()`)

Reply:
Hashables are for different reason. Hashes are not used for reproducibility purposes, and they are not returned in the R code. This single hash is needed to identify an app. At this moment it is used as "id" for `filter` argument, to check whether filter is applied to the same app. This means:
- if you have an app with dynamic data source you'd like to make sure that filters which you upload are done for this app.
- filters done for other app (possibly for other modules and data) should not be possible to upload for this app

Reply:
Should we update the comment a few lines up to be more descriptive?

Reply:
> Should we update the comment a few lines up to be more descriptive?

Yes - that would be great for a person from the outside.

> This single hash is needed to identify an app

So I was wrong with my interpretation but the problem stays the same. The way how you calculate the hash now identify whether the data loading module is the same. You can have multiple _apps_ with the same _data loading module_. 

I have other objections for other cases of that if but it's going beyond the scope of this PR so will keep quiet here.

Reply:
If app1 and app2 use the same data module, then app2 _can_ use a filter snapshot created in app1.
Note that `modules` is also hashed.

Reply:
> If app1 and app2 use the same data module, then app2 _can_ use a filter snapshot created in app1. Note that `modules` is also hashed.

So then few more comments:

1. Just to clarify as I wanted to understand this better.
IMO those two sentences conflicting with each other. If app1 and app2 uses the same data module but different analysis modules then the has won't be identical (as we include modules as well) and we cannot re-use the filter. Am I missing something here?

2. Also: are you sure?

Let's leave 1. and assume same data module and same analysis modules. Let's assume the following:
- module X that allows to load multiple csv files and apply datanames according to file names
- APP1: data module X that returns datasets A, B, C + module on "all" datanames
- APP2: data module X that returns datasets K, L, M + module on "all" datanames

So the hash here will be the same and are you sure that we can re-apply filters from APP1 to APP2? Am I right here?

I have the impression that this becomes imperfect as we introduce the concept of teal_data_module in which the same module can return different datasets. I am just trying to find a good test case.

Reply:
Ad 1. Sorry for the confusing phrasing. I did mean that in order to use an uploaded snapshot, it must come from an app that used the same data (module) **and** modules. You are right, those are not exactly different apps. Which is the point ot the hash.

Ad 2. For the data module to return different (or even differently named) datasets it would have to have either a) different server code or b) the same server code that _modifies_ data (see the new vignette for examples), where different data is supplied. The former seems safe enough. The latter, not necessarily.

Reply:
AD 1. Thanks. Now it's more clear.

AD 2. I am actually thinking about yet another possibility of dynamically named object -> things like: `x <- foo_named_list_of_df(input$foo, input$bar); list2env(x, env = current_env())`. The server code is the same but it can return different things based on user inputs. This essentially means that our identifier mechanism based on hashing of server fun body is imperfect (also taking into account use case b) you mentioned).

Reply:
Right. So in the particular case you describe snapshot compatibility is upheld. Data names may be different but the data objects themselves are the same. If you decide to change column names, _then_ you're in trouble.

The question is, is this a valid use case? Are dataset names (let alone variable names) ever left up to the app user? We don't even provide for the app user to modify variable labels (yet).

Reply:
I think it is. To me, that's one of the greatest things about teal-data-module concept. It essentially unlocks total flexibility of data names, column names, column types and everything else. We even say in the docs that (the only) limitation is that the returned object has to be of `teal_data` class. How you achieve that - it's up to the module developer. And that's great.

This feature does not fit well with our ~concept~ implementation of filters re-applicability which I am highlighting here. Now I don't know if we should invest in finding appropriate way of implement this this (here) or revisit this as a whole.